modules/axtask/src/run_queue.rs

```diff
@@ -1,72 +1,249 @@
+/// Retrieves a `'static` reference to the run queue corresponding to the given index.
+///
+/// This function asserts that the provided index is within the range of available CPUs
+/// and returns a reference to the corresponding run queue.
+///
+/// ## Arguments
+///
+/// * `index` - The index of the run queue to retrieve.
+///
+/// ## Returns
+///
+/// A reference to the `AxRunQueue` corresponding to the provided index.
+///
+/// ## Panics
+///
+/// This function will panic if the index is out of bounds.
+///
+#[inline]
+fn get_run_queue(index: usize) -> &'static mut AxRunQueue {
+    assert!(index < axconfig::SMP);
+    unsafe { RUN_QUEUES[index].assume_init_mut() }
+}
 
+/// Selects the appropriate run queue for the provided task.
+///
+/// * In a single-core system, this function always returns a reference to the global run queue.
+/// * In a multi-core system, this function selects the run queue based on the task's CPU affinity and load balance.
+///
+/// ## Arguments
+///
+/// * `task` - A reference to the task for which a run queue is being selected.
+///
+/// ## Returns
+///
+/// A reference to the selected `AxRunQueue`.
+///
+/// ## TODO
+///
+/// 1. Implement better load balancing across CPUs for more efficient task distribution.
+/// 2. Use a more generic load balancing algorithm that can be customized or replaced.
+///
+#[inline]
+pub(crate) fn select_run_queue<G: BaseGuard>(
+    #[cfg(feature = "smp")] task: AxTaskRef,
+) -> AxRunQueueRef<'static, G> {
+    #[cfg(not(feature = "smp"))]
+    {
+        // When SMP is disabled, all tasks are scheduled on the same global run queue.
+        current_run_queue()
+    }
+    #[cfg(feature = "smp")]
+    {
+        let irq_state = G::acquire();
+        // When SMP is enabled, select the run queue based on the task's CPU affinity and load balance.
+        let index = select_run_queue_index(task.cpumask());
+        AxRunQueueRef {
+            inner: get_run_queue(index),
+            state: irq_state,
+            _phantom: core::marker::PhantomData,
+        }
+    }
+}
+
+/// AxRunQueue represents a run queue for global system or a specific CPU.
 pub(crate) struct AxRunQueue {
+    /// The ID of the CPU this run queue is associated with.
+    cpu_id: usize,
+    /// The core scheduler of this run queue.
     scheduler: Scheduler,
 }
 
+pub(crate) struct AxRunQueueRef<'a, G: BaseGuard> {
+    inner: &'a mut AxRunQueue,
+    state: G::State,
+    _phantom: core::marker::PhantomData<G>,
+}
+
+impl<'a, G: BaseGuard> Drop for AxRunQueueRef<'a, G> {
+    fn drop(&mut self) {
+        G::release(self.state);
+    }
+}
+
 impl AxRunQueue {
-    pub fn new() -> SpinNoIrq<Self> {
+    pub fn new(cpu_id: usize) -> Self {
         let gc_task = TaskInner::new(gc_entry, "gc".into(), axconfig::TASK_STACK_SIZE).into_arc();
+        // gc task shoule be pinned to the current CPU.
+        gc_task.set_cpumask(CpuMask::from_usize(1 << cpu_id));
```

Comment:
The `from_usize` API is bad, change to `from_raw_bits(1 << cpu_id)` or `oneshot(cpu_id)`.

Reply:
ok